[[ ForeignCallbacks ]] Implement basic version of C callable LCB clos…

runrevmark opened the pull request:
…ures.

Passing LCB handlers to foreign functions can now be done easily.

A 'foreign handler type' should be declared which describes the signature of the handler and the appropriate parameter of a foreign handler definition should use this type.

When the VM executes a foreign handler, it will automatically bridge any foreign handler type parameters to C function pointers (by creating a libffi closure).

The lifetime of the C function pointers generated in this way are the same as the lifetime of the instance which created them. i.e. For library modules they will be around for as long as the module is loaded, for widgets they will be around for as long as the widget is.

Example:

```
public foreign handler type MCProperListApplyCallback(in pContext as optional Pointer, in pElement as optional any) returns CBool

foreign handler MCProperListApply(in pList as List, in pCallback as MCProperListApplyCallback, in pContext as optional Pointer) returns CBool binds to "<builtin>"
```

Here, any compatible LCB handler passed to argument 2 of MCProperListApply will cause an automatic generation of a C function pointer which the C code can use to invoke the LCB handler.

runrevmark committed:
[[ ForeignCallbacks ]] Implement basic version of C callable LCB closures.

Code to compute the FFI layout type for handler types has been moved into MCHandlerTypeInfo.

Additionally, an MCHandlerGetFunctionPtr API has been added which attempts to create and then
return a C function pointer which will execute the HandlerRef.

In order to ensure that the Handler value and its closure are not released it is important to
keep the handler value in a variable for the lifetime the function pointer is required.

peter-b commented on libfoundation/src/foundation-handler.cpp:
```diff
@@ -90,10 +93,141 @@ const MCHandlerCallbacks *MCHandlerGetCallbacks(MCHandlerRef self)
+        t_type = MCHandlerTypeInfoGetParameterType(t_signature, t_arg_index);
+        
+        if (t_mode != kMCHandlerTypeFieldModeIn)
+            abort();
```
`MCAssert()` here?

runrevmark replied:
I've made this throw an Unimplemented error - similarly if more than 16 args are attempted to be used.

peter-b commented on libfoundation/src/foundation-handler.cpp:
```diff
@@ -90,10 +93,141 @@ const MCHandlerCallbacks *MCHandlerGetCallbacks(MCHandlerRef self)
+                if (t_descriptor -> bridgetype != kMCNullTypeInfo)
+                {
+                    if (!t_descriptor -> doimport(args[t_arg_index], false, t_value_args[t_arg_index]))
+                        goto error_exit;
```
Do you need to `MCErrorThrow...` here?

peter-b commented on libfoundation/src/foundation-handler.cpp:
```diff
@@ -90,10 +93,141 @@ const MCHandlerCallbacks *MCHandlerGetCallbacks(MCHandlerRef self)
+        MCResolvedTypeInfo t_resolved_type;
+        if (!MCTypeInfoResolve(t_type, t_resolved_type))
+        {
+            MCErrorThrowGeneric(nil);
```
Error message is missing.

peter-b replied:
It looks like throwing an error on failure of `MCTypeInfoResolve()` is a fairly common thing to do, so maybe `MCTypeInfoResolve()` should itself throw an error on failure.

runrevmark replied:
I've refactored the UnboundTypeError message from libscript into libfoundation and added a throw method for it. I'm not sure MCTypeInfoResolve should necessarily throw - there might be current contexts where that wouldn't be correct.

To be fair, the type resolution needs a little bit of a rethink. There should be no need to resolve the types really as everything should be patched up as extensions load (but that isn't the case at the moment).

peter-b commented on libfoundation/src/foundation-handler.cpp:
```diff
@@ -90,10 +93,141 @@ const MCHandlerCallbacks *MCHandlerGetCallbacks(MCHandlerRef self)
+        MCErrorThrowGeneric(nil);
+        goto error_exit;
+    }
+    
```
What about validating the return typeinfo _before_ invoking the handler?

livecodestephen replied:
I'm guessing that line 164 shouldn't be identical to line 161.

runrevmark replied:
Indeed :)

peter-b commented on libfoundation/src/foundation-handler.cpp:
```diff
@@ -90,10 +93,141 @@ const MCHandlerCallbacks *MCHandlerGetCallbacks(MCHandlerRef self)
+        MCValueRelease(t_value_result);
+    for(uindex_t i = 0; i < t_arg_index; i++)
+        MCValueRelease(t_value_args[i]);
+}
```
Is it possible to use `MCAutoValueRef` and `MCAutoValueRefArray` to avoid explicit cleanup rules?

runrevmark replied:
I didn't do that at the moment as it uses a fixed 16 element array for args. That being said, i think there's a memory leak here if things succeed - as the releases are only done on the error path.

peter-b commented on libfoundation/src/foundation-handler.cpp:
```diff
@@ -90,10 +93,141 @@ const MCHandlerCallbacks *MCHandlerGetCallbacks(MCHandlerRef self)
+    {
+        ffi_closure_free(self -> closure);
+        self -> closure = nil;
+        return MCErrorThrowGeneric(nil);
```
Missing error message.

runrevmark replied:
I'm not sure that can ever be triggered - ffi_prep_closure_loc can only fail if the ABI is unknown (it seems). Perhaps an MCErrorThrowUnreachable()?

peter-b commented on libfoundation/src/foundation-typeinfo.cpp:
```diff
@@ -771,6 +774,103 @@ MCTypeInfoRef MCHandlerTypeInfoGetParameterType(MCTypeInfoRef unresolved_self, u
+            MCResolvedTypeInfo t_resolved_type;
+            if (!MCTypeInfoResolve(t_type, t_resolved_type))
+            {
+                t_success = false;
```
You could just `return MCErrorThrow...` from here instead of using the `t_success` variable.

peter-b commented on libfoundation/src/foundation-typeinfo.cpp:
```diff
@@ -771,6 +774,103 @@ MCTypeInfoRef MCHandlerTypeInfoGetParameterType(MCTypeInfoRef unresolved_self, u
+        }
+
+        if (!t_success)
+            return MCErrorThrowGeneric(nil);
```
Missing error message.

peter-b commented:
I'm really looking forward to trying this out. :+1: libcurl bindings here we come!

runrevmark commented:
It does need a bit more work - it's the 'minimum necessary' that Ian needed to finish off the browser-related work he's doing.

livecodeian commented:
This fails when the input is a native foreign handler - MCHandlerGetFunctionPtr is passed a null MCHandlerRef, resulting in a crash.

livecode-vulcan commented:
📝 Hi @runrevmark, I haven't been able to verify that you've signed our [LiveCode Contributor's Agreement](https://livecode.com/account/developer/contribute).

If you have previously signed the Contributor's Agreement, I may not be able to detect it because you haven't [linked your GitHub account to your LiveCode account](https://livecode.com/account/developer/github).

Please see the [information for contributors](https://github.com/runrev/livecode/blob/develop/HACKING.md) for more information.

runrevmark committed:
Merge branch 'develop' into lcb-native_callback

runrevmark committed:
[[ Foundation ]] Added UnboundType and Unimplemented errors with appropriate Throw methods.

runrevmark committed:
[[ ForeignCallbacks ]] Minor cleanup.

livecode-vulcan commented:
📝 Hi @runrevmark, I haven't been able to verify that you've signed our [LiveCode Contributor's Agreement](https://livecode.com/account/developer/contribute).

If you have previously signed the Contributor's Agreement, I may not be able to detect it because you haven't [linked your GitHub account to your LiveCode account](https://livecode.com/account/developer/github).

Please see the [information for contributors](https://github.com/runrev/livecode/blob/develop/HACKING.md) for more information.

peter-b commented:
@livecode-vulcan review ok 92ef50a

livecode-vulcan commented:
💙 review by @peter-b ok 92ef50a0ba1a9a3cec22e3fdd0ae48e4a6e990c2

runrevmark committed:
[[ ForeignCallbacks ]] Minor cleanup.

livecode-vulcan commented:
😎 test success 92ef50a0ba1a9a3cec22e3fdd0ae48e4a6e990c2
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/226)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/225)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/226)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/221)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/228)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/224)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/229)

peter-b commented:
@runrevmark I'm pretty certain this does still need release notes...

runrevmark commented:
I think I removed labels from the wrong PR...

runrevmark committed:
[[ Foundation ]] Fix memory management issues in closure callback.

runrevmark committed:
[[ ForeignCallbacks ]] Add initial (simple) unit test.

runrevmark commented:
In regards to docs and release notes - this can't really be a 'supported public feature' at the moment. It depends on using an engine method to get the function ptr and you need to understand the lifetime model of the function ptr to use it. It's 'good enough' for careful use in the browser module, but needs a little more work to make it a fully-fledged usable feature.

runrevmark commented:
Okay having pondered this then the mapping to the C function pointer is an FFI related matter. However, the lifetime issue is one which needs thought.

Obviously we cannot make passing function ptrs to native code completely safe, but what we can do is add support to ensure the developer doesn't 'do the wrong thing'. Given any request for a function-ptr then it always has an associated required-lifetime. For example, when passing a callback to MCProperListApply, the function-ptr only needs to exist for as long as the call to that method; in contrast, for a widget hooking into a native 'thing' the lifetime would be more appropriate to be the lifetime of the widget instance.

The general mechanism (allowing function-ptrs to be derived from arbitrary closures) could be done using the idea of 'pinning'. In order for an LCB HandlerRef to be passed to a native function, the developer would need to 'pin' and 'unpin' the handler appropriately. This would be a pain for all cases however.

To solve this, what we could do is add annotations to FFI definitions taking function pointers which indicate what their lifetime requirements are.

For example, for MCProperListApply, the function ptr would be annotated as 'restricted to call' - meaning that the function ptr only needs to be pinned either side... Something the VM could do automagically.

Similarly, for a callback being passed to an native 'thing' which is created and destroyed in a widget's Create/Destroy handlers, the function ptr would be annotated as 'restricted to instance'. (This is the same for normal modules - its just that a normal module's instance exists for the 'lifetime of the app' in theory anyway). Implementing this would require that a ScriptInstance holds a list of HandlerRefs with non-nil C function ptrs, and then it could clean them up when it disappears.

So, we probably need a 'pinning' idea; then augment the VM / FFI definitions to enable automatic pinning by the VM. Finally, the VM would throw an error if you attempt to pass a handler to a native function which is unpinned, or cannot be automatically pinned (because the VM doesn't know the lifetime).

peter-b commented:
@runrevmark What about making the lifetime a property of the foreign handler **type** declaration, rather than of the foreign handler declaration?  What about foreign handlers that accept multiple function callbacks with different lifetime requirements?  (Just playing Devil's Advocate here).

runrevmark commented:
@peter-b: Is that any better? After all you might have a handler type which could require different lifetimes, just as much as you might have a foreign handler which bestows different lifetimes upon its handler arguments.

My thought was that the lifetime is a property of the parameter to the function - the processing of the lifetime occurs at the point of the bridge from LCB to FFI (a bit like the bridging from Number to int32_t, for example).

Also, if it is an attribute of the handler type, that might suggest it has some effect on the value of the handler type in general when in actual fact the lifetime (automatic pinning) support only has relevance when it passes to a foreign handler.

At the end of the day, you'll either end up declaring multiple handler types, or multiple bindings to the same underlying foreign function in a situation such as this.

(I should say that I was proposing an annotation on specific handler type parameters in foreign handler definitions, rather than one annotation for all handler parameters).

peter-b commented:
@runrevmark Thinking about this a bit further... for now, this is an obviously super useful capability to have.  I think that it deserves to be documented and appear in release notes, because it's there.  The whole of LiveCode Builder is experimental, and subject to bugs/changes, and this is (clearly) no exception.  We can make it clear what the current lifetime requirements are, and warn of dire consequences for violating them.  The whole of LiveCode Builder's FFI is fairly "_caveat_ developer" at the moment, so one more (highly useful!) example won't be a disaster!

Yes, it makes sense to put the lifetime annotation per-parameter in the `foreign handler` declaration, I think.

peter-b commented:
@runrevmark I missed out the end of my last sentence:

...but since the there's only one lifetime available at the moment I don't see any problem with leaving it implicit.

runrevmark commented:
@peter-b: Indeed, actually you make a good point. We don't actually need to do too much to make this 'nicer'. The first thing is to add a 'foreign handler type' - so the FFI bridge knows when a handler wants an MCHandlerRef or a C-function-ptr; the second is to cache all HandlerRef's that have been created for a given instance, in the instance. The only available lifetime is then 'lifetime of instance' (whether that be a widget, or a library).

mwieder commented:
OMG I totally missed this. libcurl bindings !
I think if you start worrying about lifetimes too much you're going to end up with garbage collection

runrevmark committed:
[[ Foreign Callbacks ]] Bridging and lifetime management.

There is a new LCB type 'foreign handler type'. Handler types marked as foreign will automatically bridge
to C function pointers when used as a parameter type for a foreign handler.

Additionally, the lifetime of C function pointers generated in this way is guaranteed to be the length
of the existance of the instance they are attached to. i.e. When a handler identifier is evaluated, a
handler-ref is cached against the module instance; thus if the handler-ref is subsequently bridged to a
C function ptr, that handler-ref owns the C function ptr. When a module instance is finally destroyed, all
handler-refs and thus any C function pointers they own are then destroyed.

runrevmark commented:
@livecodefraser That's a good point - I'll move it into the flags word.

runrevmark committed:
[[ ForeignCallbacks ]] Moved is_foreign bool to MCValueRef flags word to save space.

runrevmark committed:
[[ ForeignCallbacks ]] Updated LCB spec.

runrevmark committed:
[[ ForeignCallbacks ]] Added release note.

livecodefraser commented:
@livecode-vulcan review okay 828ae61

runrevmark committed:
[[ ForeignCallbacks ]] Added native callback related tests.

livecodefraser commented:
@livecode-vulcan review okay c4ddfe8

livecode-vulcan commented:
💙 review by @livecodefraser ok c4ddfe84f076495a3baf56fb97f29161e37d8517

livecode-vulcan commented:
😎 test success c4ddfe84f076495a3baf56fb97f29161e37d8517
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/274)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/272)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/273)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/272)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/277)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/269)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/280)

livecodeali commented on libscript/src/script-instance.cpp:
```diff
@@ -2472,6 +2387,112 @@ bool MCScriptCallHandlerOfInstanceInternal(MCScriptInstanceRef self, MCScriptHan
+        
+        // The context struct is 'moved' into the handlerref.
+        __MCScriptHandlerContext t_context;
+        t_context . instance = MCScriptRetainInstance(p_instance);
```
This retain is preventing script instances with handler objects being destroyed. I think MCScriptReleaseInstance should go through its handler array releasing any instances which have been retiained in the context.
